### PR TITLE
Fix gather_for_metrics

### PR DESCRIPTION
**Pull request by muellerzr**
Solves https://github.com/huggingface/accelerate/issues/575 by completely changing how we check for the number of samples to drop. Instead uses modulo to learn how many extra samples there are initially, allowing us to simplify the logic.

### CODE REVIEW

File: src/accelerate/accelerator.py [hunk 1]
```diff
@@ -934,25 +934,28 @@ def gather(self, tensor):
...
                 The tensors for calculating metrics across all processes.
-            dataloader (`torch.utils.data.DataLoader`):
-                A dataloader prepared with `Accelerator.prepare`
         """
         tensor = self.gather(tensor)
         if self.use_distributed:
+            if self.gradient_state.remainder == -1:
+                logger.warn(
+                    "The used dataset had no length, returning gathered tensors. You should drop the remainder yourself."
+                )
```
Comment (sgugger) [inline src/accelerate/accelerator.py:951]: I think this can be an info, no need for a warning here.

File: src/accelerate/data_loader.py [hunk 2]
```diff
@@ -332,7 +336,11 @@ def __iter__(self):
         if self.rng_types is not None:
             synchronize_rng_states(self.rng_types, self.generator)
         self.gradient_state._set_end_of_dataloader(False)
-        self.gradient_state._set_samples_seen(0)
+        try:
+            length = getattr(self.dataset, "total_dataset_length", len(self.dataset))
+            self.gradient_state._set_remainder(length % self.batch_size)
+        except:
```
Comment (sgugger) [inline src/accelerate/data_loader.py:342]: There should be an exception ehre :-)

File: src/accelerate/test_utils/scripts/test_metrics.py [hunk 3]
```diff
@@ -22,50 +22,63 @@
...
+    inps, targs = torch.cat(inps), torch.cat(targs)
+    return inps, targs
+
+
+def test_torch_metrics(accelerator: Accelerator, num_samples=82):
+    model, ddp_model, dataloader = get_setup(accelerator, num_samples)
+    inps, targs = generate_predictions(ddp_model, dataloader, accelerator)
+    assert (
+        len(inps) == num_samples
+    ), f"Unexpected number of inputs:\n    Expected: {num_samples}\n    Actual: {len(inps)}"
```
Comment (sgugger) [inline src/accelerate/test_utils/scripts/test_metrics.py:58]: Nice new test!

**Review by sgugger — APPROVED** [review]
Very nice fix!

**Review by sgugger — APPROVED** [review]

File: src/accelerate/data_loader.py [hunk 4]
```diff
@@ -453,7 +463,11 @@ def _fetch_batches(self, iterator):
 
     def __iter__(self):
         self.gradient_state._set_end_of_dataloader(False)
-        self.gradient_state._set_samples_seen(0)
+        try:
+            length = getattr(self.dataset, "total_dataset_length", len(self.dataset))
+            self.gradient_state._set_remainder(length % self.batch_size)
```
Comment (pacman100) [inline src/accelerate/data_loader.py:468]:
```suggestion
self.gradient_state._set_remainder(length % self.total_batch_size)
```

**Review by pacman100 — COMMENTED** [review]
Great fix, simple and intuitive 😄. I think the below change is required in `DataLoaderDispatcher`.

File: src/accelerate/data_loader.py (hunk 4 above)
Comment (pacman100) [reply src/accelerate/data_loader.py:468]: After the above change or without it, `test_metrics.py` fails when `dispatch_batches=True` when the `num_samples % num_processes != 0` , e.g., num_samples=97 and num_processes=2. Below is the modified test with print statements. I think, `IterableDatasetShard` should be used in conjunction with  `dispatch_batches=True` for expected behaviour or `DataLoaderDispatcher` needs to be fixed to handle normal dataloaders. 

Screenshot 2022-07-28 at 10 54 03 AM

Comment (muellerzr) [reply src/accelerate/data_loader.py:468]: Thanks, will change the test case slightly for this and resolve today!

**Pull request merged**